src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs

```diff
@@ -4562,8 +4570,124 @@ public void TestSemicolonAfterAnonymousTypeMemberComma()
             Assert.NotEqual(SyntaxKind.None, ds.Declaration.Variables[0].Initializer.EqualsToken.Kind());
             Assert.NotNull(ds.Declaration.Variables[0].Initializer.Value);
             Assert.Equal(SyntaxKind.AnonymousObjectCreationExpression, ds.Declaration.Variables[0].Initializer.Value.Kind());
-            Assert.Equal(1, file.Errors().Length);
-            Assert.Equal((int)ErrorCode.ERR_RbraceExpected, file.Errors()[0].Code);
+            Assert.Equal(3, file.Errors().Length);
+            Assert.Equal((int)ErrorCode.ERR_IdentifierExpected, file.Errors()[0].Code);
+            Assert.Equal((int)ErrorCode.ERR_SemicolonExpected, file.Errors()[1].Code);
+            Assert.Equal((int)ErrorCode.ERR_RbraceExpected, file.Errors()[2].Code);
+        }
+
+        [Fact]
+        public void SemicolonSeparatorInArrayInitializer()
+        {
+            var source = """
+        class Program
+        {
+            static void Main()
+            {
+                F();
+                var a = new[] { 1; 2 };
+            }
+            static object F()
+            {
+                return null;
+            }
+        }
+        """;
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (6,26): error CS1003: Syntax error, ',' expected
+                //         var a = new[] { 1; 2 };
+                Diagnostic(ErrorCode.ERR_SyntaxError, ";").WithArguments(",").WithLocation(6, 26),
+                // (6,28): error CS1003: Syntax error, ',' expected
+                //         var a = new[] { 1; 2 };
+                Diagnostic(ErrorCode.ERR_SyntaxError, "2").WithArguments(",").WithLocation(6, 28));
```

Comment:
this doesn't seem correct.  if we are treating `;` as the separator, we should def get the first error, but not the second.  the '2' should be fine now as it's an acceptable expression in the initializer.

We should also dump the tree here (see the UsingTree function that lots of tests use).  So we can show that the error recovery worked.

Reply:
this is currently a blocking concern for me as it indicates something is not working here as i would expect based on the goals of this PR.

Reply:
I've tried a different approach and managed to achieve that the second error does not show up. I'll push the commits hopefully tomorrow. Then the changes should be more in line with what you would expect.